### PR TITLE
Add shopping cart model

### PR DESCRIPTION
**Pull request by JohnGrey17**

### CODE REVIEW

**Review by HAIOVYI — APPROVED** [review]
GJ

File: src/main/java/org/example/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                                                 @RequestBody @Valid
+                                                 CartItemUpdateDto updateDto) {
+        return shoppingCartService.updateCartItemById(cartItemId, updateDto);
+    }
+
+    @DeleteMapping("/cart-items/{cartItemId}")
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    @Operation(summary = "Delete Cart Item", description = "Delete a cart item "
+            + "with the given ID.")
+    public void deleteCartItemById(@PathVariable Long cartItemId) {
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/controller/ShoppingCartController.java:68]:
```suggestion
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
    public void deleteCartItemById(@PathVariable Long cartItemId) {
```

File: src/main/java/org/example/bookstore/dto/cartitemdto/CartItemUpdateDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+public class CartItemUpdateDto {
+    private String bookTitle;
+    @Min(1)
+    private Integer quantity;
+}
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/dto/cartitemdto/CartItemUpdateDto.java]: image

File: src/main/java/org/example/bookstore/mapper/CartItemMapper.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.example.bookstore.model.Book;
+import org.example.bookstore.model.CartItem;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.Named;
+
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+
+    @Mapping(target = "bookId", source = "book", qualifiedByName = "bookIdByBook")
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/mapper/CartItemMapper.java]:
```suggestion
    @Mapping(target = "bookId", source = "book.id")
```
this should work

File: src/main/java/org/example/bookstore/mapper/CartItemMapper.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    @Mapping(target = "bookTitle", source = "book", qualifiedByName = "getTitleByBook")
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/mapper/CartItemMapper.java]:
```suggestion
    @Mapping(target = "bookTitle", source = "book.title")
```

File: src/main/java/org/example/bookstore/model/Book.java [hunk 5]
```diff
@@ -41,7 +41,7 @@ public class Book {
     private String description;
     @Column (name = "cover_image")
     private String coverImage;
-    @ManyToMany(fetch = FetchType.LAZY)
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/model/Book.java]: EAGER is a bad solution. Let's remember how to use LAZY and save its efficiency

File: src/main/java/org/example/bookstore/model/Book.java [hunk 6]
```diff
@@ -41,7 +41,7 @@ public class Book {
     private String description;
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/model/Book.java]: please add `@EqualsAndHashCode` to your entities

File: src/main/java/org/example/bookstore/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+@SQLRestriction(value = "is_deleted=false")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

File: src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                        + "not found for id: "
+                        + requestDto.getBookId())));
+        cartItem.setShoppingCart(shoppingCart);
+        shoppingCart.getCartItems().add(cartItem);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getUserShoppingCartById(Long userId, Pageable pageable) {
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java]: I think you don't need `Pageable` here

File: src/main/java/org/example/bookstore/repository/shoppingcart/ShoppingCartRepository.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart,Long> {
+
+    @Query("SELECT sc FROM ShoppingCart sc LEFT JOIN FETCH sc.cartItems WHERE sc.user.id = :userId")
+    ShoppingCart findShoppingCartByUserId(@Param("userId") Long userId);
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/repository/shoppingcart/ShoppingCartRepository.java]: should be `Optional<ShoppingCart>`

File: src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        ShoppingCart shoppingCartByUserId = shoppingCartRepository.findShoppingCartByUserId(userId);
+        return shoppingCartMapper.toDto(shoppingCartByUserId);
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java]: there should be a `null` check

File: src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto addBookToCartItem(Long userId, CartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(userId);
+        if (shoppingCart == null) {
+            throw new DataIntegrityViolationException("Shopping cart "
+                    + "not found for user id: " + userId);
+        }
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java]: you can move this logic into a private method and reuse in `getUserShoppingCartById`

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
Great job! Left some comments

File: src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 12]
```diff
@@ -39,12 +38,9 @@ public void createNewShoppingCart(User user) {
...
     public ShoppingCartResponseDto addBookToCartItem(Long userId, CartItemRequestDto requestDto) {
-        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(userId);
-        if (shoppingCart == null) {
-            throw new DataIntegrityViolationException("Shopping cart "
-                    + "not found for user id: " + userId);
-        }
-
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("can`t find shopping "
+                        + "can`t by id: " + userId));
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:43]: this code is duplicated

File: src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 13]
```diff
@@ -58,8 +54,11 @@ public ShoppingCartResponseDto addBookToCartItem(Long userId, CartItemRequestDto
     }
 
     @Override
-    public ShoppingCartResponseDto getUserShoppingCartById(Long userId, Pageable pageable) {
-        ShoppingCart shoppingCartByUserId = shoppingCartRepository.findShoppingCartByUserId(userId);
+    @Transactional
+    public ShoppingCartResponseDto getUserShoppingCartById(Long userId) {
+        ShoppingCart shoppingCartByUserId = shoppingCartRepository.findShoppingCartByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("can`t find shoppingCart by id: "
+                        + userId));
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/bookstore/service/shoppingcart/ShoppingCartServiceImpl.java:61]: this code is duplicated

**Review by yevheniisynytsia — APPROVED** [review]

**Pull request closed without merging**